### PR TITLE
[fix] Dropout2d-3d no-batch-dim

### LINKED ISSUE #69801: Dropout2d doesn't drop channels for (C, H, W)
### 🐛 Describe the bug

[Docs](https://pytorch.org/docs/1.10.0/generated/torch.nn.Dropout2d.html?highlight=dropout2d#torch.nn.Dropout2d) read that `C` is dropped, which does not occur for `(C, H, W)` 

```python
import torch
torch.manual_seed(0)
ipt = torch.ones((2, 3, 4))
print(torch.nn.Dropout2d(p=0.5)(ipt))
```
```
tensor([[[2., 2., 2., 2.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],
        [[2., 2., 2., 2.],
         [2., 2., 2., 2.],
         [0., 0., 0., 0.]]])
```
but does for `(N, C, H, W)`
```python
torch.manual_seed(0)
print(torch.nn.Dropout2d(p=0.5)(ipt[None]))
```
```
tensor([[[[2., 2., 2., 2.],
          [2., 2., 2., 2.],
          [2., 2., 2., 2.]],
         [[0., 0., 0., 0.],
          [0., 0., 0., 0.],
          [0., 0., 0., 0.]]]])
```
It appears `Dropout2d` is instead implementing `Dropout1d` for `(N, C, T)`.

### Versions

1.10.0 via Anaconda; Windows 10

cc @ezyang @gchanan @zou3519 @bdhirsh @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/69801

TODO:
* [x] Update C++ API

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -13520,6 +13520,15 @@ def test_Dropout(self, device):
             input = input.bfloat16()
             self._test_dropout(nn.Dropout, device, input)
 
+    def _test_dropoutNd_no_batch(self, dropout, input):
+        with freeze_rng_state():
+            res_no_batch = dropout(input)
+
+        with freeze_rng_state():
+            res_batched = dropout(input.unsqueeze(0)).squeeze(0)
```
Comment (jbschlosser) [inline test/test_nn.py:13528]: tiny nit: I think you could do these both under the same `with freeze_rng_state():`

**Review by jbschlosser — COMMENTED** [review]
Thanks for the fix!

FYI I think the failing checks are happening because the OpInfo for `nn.functional.dropout2d` is wrong - looks like it's generating inputs with dim in range [2, 4] instead of [3, 4].

File: test/test_nn.py (hunk 1 above)
Comment (kshitij12345) [reply test/test_nn.py:13528]: They should be different. Looking at the source, we reset the seed after exiting the context.
https://github.com/pytorch/pytorch/blob/408283319ab6b0eefcc756b6affa3fe7d59db9db/torch/testing/_internal/common_utils.py#L1128-L1144

Also I verified with the following snippet
```python
import torch

from torch.testing._internal.common_utils import freeze_rng_state

with freeze_rng_state():
    print(torch.randn(2))

with freeze_rng_state():
    print(torch.randn(2))
    print("*"*10)

# In same context
with freeze_rng_state():
    print(torch.randn(2))
    print(torch.randn(2))
    print("*"*10)

```

Output
```
tensor([0.0461, 0.4024])
tensor([0.0461, 0.4024])
**********
tensor([0.0461, 0.4024])
tensor([-1.0115,  0.2167])
**********
```

**Comment by kshitij12345** [comment]
Looks like JIT tests are also failing. Will have a look.

**Comment by kshitij12345** [comment]
@jbschlosser I couldn't find a test for `Dropout2d` and `Dropout3d` where we check that channel is being zeroed, not sure as to how we can test that.

File: test/test_nn.py (hunk 1 above)
Comment (jbschlosser) [reply test/test_nn.py]: Awesome, thanks for clearing that up!

**Comment by jbschlosser** [comment]
> @jbschlosser I couldn't find a test for `Dropout2d` and `Dropout3d` where we check that channel is being zeroed, not sure as to how we can test that.

One option is to input a tensor that is known to be fully non-zero and check that the output only has entire channels zeroed out (i.e. anywhere a zero is present, the entire channel should also be zero).

File: test/test_nn.py [hunk 2]
```diff
@@ -13578,6 +13578,29 @@ def test_Dropout(self, device):
...
+        shape = input.shape
+        B = shape[0]
+        C = shape[1]
+        channel_numel = torch.tensor(shape[2:]).prod()
+        result = dropout(input)
+
+        for b, c in product(range(B), range(C)):
+            self.assertTrue(result[b, c].count_nonzero() in (0, channel_numel))
+
+    @expectedFailureXLA  # seems like freeze_rng_state is not honoured by XLA
```
Comment (jbschlosser) [inline test/test_nn.py:13603]: boo :(

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -14601,7 +14601,8 @@ def ref_pairwise_distance(input1, input2):
         supports_forward_ad=True,
         supports_fwgrad_bwgrad=True,
         supports_out=False,
-        sample_inputs_func=partial(sample_inputs_dropout, train=True, min_input_dim=2),
+        # As per the docs, valid input dims are (4, 5)
+        sample_inputs_func=partial(sample_inputs_dropout, train=True, valid_input_dim=(4, 5)),
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:14605]: This is OpInfo entry for `feature_alpha_dropout`.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -14575,7 +14574,8 @@ def ref_pairwise_distance(input1, input2):
         supports_forward_ad=True,
         supports_fwgrad_bwgrad=True,
         supports_out=False,
-        sample_inputs_func=partial(sample_inputs_dropout, min_input_dim=2),
+        # As per the docs, valid input dims are (3, 4)
+        sample_inputs_func=partial(sample_inputs_dropout, valid_input_dim=(3, 4)),
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:14578]: This is OpInfo entry for `dropout2d`.

File: test/cpp/api/functional.cpp [hunk 5]
```diff
@@ -2574,35 +2574,31 @@ TEST_F(FunctionalTest, Dropout) {
...
 TEST_F(FunctionalTest, Dropout2d) {
-  auto input = torch::randn({50, 100});
+  auto input = torch::randn({2, 2, 50, 100});
   auto input_mean = input.mean();
   auto input_std = input.std();
 
   for (const auto rate : {0.2, 0.5, 0.8}) {
     auto output = F::dropout2d(input, F::Dropout2dFuncOptions().p(rate));
     ASSERT_TRUE(torch::allclose(input_mean, output.mean(), 0.01, 0.05));
-    ASSERT_TRUE((input_std <= output.std()).all().item<bool>());
```
Comment (kshitij12345) [inline test/cpp/api/functional.cpp:2584]: Not necessarily true.
```python
import torch
import torch.nn as nn

from itertools import product
p = 0.5
m = nn.Dropout2d(p=0.5)
while True:
    inp = torch.randn(2, 2, 2, 1)
    output = m(inp)
    if (output.std() < inp.std()):
        print(output.std())
        print(inp.std())
        print(output)
        print(inp)
        break

```

Output:
```
tensor(1.0140) # out std
tensor(1.4745) # inp std
tensor([[[[ 0.0000], # out tensor
          [-0.0000]],

         [[ 0.0000],
          [-0.0000]]],


        [[[-2.8121],
          [ 0.2904]],

         [[ 0.0000],
          [ 0.0000]]]])
tensor([[[[ 2.6668], # inp tensor
          [-1.3751]],

         [[ 0.5103],
          [-0.2788]]],


        [[[-1.4061],
          [ 0.1452]],

         [[ 1.1308],
          [ 2.0132]]]])
```

Comment (jbschlosser) [reply test/cpp/api/functional.cpp:2584]: cool I agree - nice catch!

File: torch/csrc/api/include/torch/nn/functional/dropout.h [hunk 6]
```diff
@@ -46,16 +46,47 @@ inline Tensor dropout(Tensor input,
...
-    return torch::feature_dropout_(input, p, training);
+    result = torch::feature_dropout_(input, p, training);
   } else {
-    return torch::feature_dropout(input, p, training);
+    result = torch::feature_dropout(input, p, training);
   }
+
+  if (!is_batched) {
+    if (inplace) {
+      result = result.squeeze_(0);
```
Comment (jbschlosser) [inline torch/csrc/api/include/torch/nn/functional/dropout.h:80]: shouldn't this just be `result.squeeze_(0)`?

Edit: nvm this is okay :)

File: torch/nn/functional.py [hunk 7]
```diff
@@ -1303,7 +1303,20 @@ def dropout2d(input: Tensor, p: float = 0.5, training: bool = True, inplace: boo
...
     if p < 0.0 or p > 1.0:
         raise ValueError("dropout probability has to be between 0 and 1, " "but got {}".format(p))
-    return _VF.feature_dropout_(input, p, training) if inplace else _VF.feature_dropout(input, p, training)
+    inp_dim = input.dim()
+    msg = f"dropout2d: Input should be 3-D (unbatched) or 4-D (batched) but received {inp_dim}-D tensor"
+    assert inp_dim in (3, 4), msg
+
+    is_batched = inp_dim == 4
+    if not is_batched:
+        input = input.unsqueeze_(0) if inplace else input.unsqueeze(0)
```
Comment (jbschlosser) [inline torch/nn/functional.py:1312]: I guess in-place ops also return self so this is okay?

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_methods_invocations.py:14605]: This is good for the purposes of this PR, but eventually `FeatureAlphaDropout` should be documented and tested to work on dims besides 4 and 5 (see https://github.com/pytorch/pytorch/issues/69960#issuecomment-995169300)

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks for the fix :)

**Comment by jbschlosser** [comment]
@kshitij12345 Sorry for the delay on this - some internal issues came up. Do you mind rebasing?

**Comment by jbschlosser** [comment]
Hey @kshitij12345, FYI there's a whole bunch of internal breakage from this change. It seems there are people using dropout2d with 2D inputs expecting normal dropout to occur, which does happen to be the case. There may also be people passing 3D inputs expecting something, and redefining 3D to mean no-batch-dim input would silently break them, which is worse than loudly breaking them. So I propose a tweak to this PR to handle this case:
1. Start by restricting input to 4D (N, C, H, W) for Dropout2d and throwing an error for other input dims. Change the docs back to indicate that this is the only supported shape. Do this for a full release to help remove invalid usages of 3D inputs with dropout2d.
2. After a release cycle, we can add no-batch-dim support by allowing 3D. This should help avoid the silent breakage.

(do the analogous thing for Dropout3d / 5D input as well)

wdyt about this?

**Comment by kshitij12345** [comment]
Wouldn't this be weird? The support for no-batch-dim for these two modules was added in `1.10`. We remove it in `1.11` and bring it back with new semantics in `1.12`?

https://pytorch.org/docs/1.10.0/generated/torch.nn.Dropout2d.html?highlight=dropout2d#torch.nn.Dropout2d
https://pytorch.org/docs/1.10.0/generated/torch.nn.Dropout3d.html?highlight=dropout3d#torch.nn.Dropout3d

**Comment by kshitij12345** [comment]
@jbschlosser as discussed offline, I have updated the PR to be less strict.

File: torch/nn/functional.py [hunk 8]
```diff
@@ -1328,13 +1345,27 @@ def dropout3d(input: Tensor, p: float = 0.5, training: bool = True, inplace: boo
...
         return handle_torch_function(dropout3d, (input,), input, p=p, training=training, inplace=inplace)
     if p < 0.0 or p > 1.0:
         raise ValueError("dropout probability has to be between 0 and 1, " "but got {}".format(p))
-    return _VF.feature_dropout_(input, p, training) if inplace else _VF.feature_dropout(input, p, training)
+    inp_dim = input.dim()
+    msg = f"dropout3d: Input should be 4-D (unbatched) or 5-D (batched) but received {inp_dim}-D tensor"
+    assert inp_dim in (3, 4, 5), msg
+    if inp_dim == 3:
+        warn_msg = ("dropout3d: 3-D input is being deprecated for Dropout3d."
+                    " Look at the docs for correct input shapes and usage.")
```
Comment (jbschlosser) [inline torch/nn/functional.py:1357]: nit: deprecation should be actionable to remove the warning - in this case, switching to `F.dropout` is the way to go:
```
Received a 3D input to dropout3d, which is deprecated and will result in an error in a future release. To retain the behavior and silence this warning, please use dropout instead. Note that dropout3d exists to provide channel-wise dropout on inputs with 3 spatial dimensions, a channel dimension, and an optional batch dimension (i.e. 4D or 5D inputs).
```

File: torch/nn/functional.py [hunk 9]
```diff
@@ -1310,7 +1310,24 @@ def dropout2d(input: Tensor, p: float = 0.5, training: bool = True, inplace: boo
         return handle_torch_function(dropout2d, (input,), input, p=p, training=training, inplace=inplace)
     if p < 0.0 or p > 1.0:
         raise ValueError("dropout probability has to be between 0 and 1, " "but got {}".format(p))
-    return _VF.feature_dropout_(input, p, training) if inplace else _VF.feature_dropout(input, p, training)
+    inp_dim = input.dim()
+    msg = f"dropout2d: Input should be 3-D (unbatched) or 4-D (batched) but received {inp_dim}-D tensor"
+    assert inp_dim in (2, 3, 4), msg
+    if inp_dim == 2:
+        warn_msg = ("dropout2d: 2-D input is being deprecated for Dropout2d."
+                    " Look at the docs for correct input shapes and usage.")
```
Comment (jbschlosser) [inline torch/nn/functional.py:1318]: Wording suggestion as below:
```
Received a 2D input to dropout2d, which is deprecated and will result in an error in a future release. To retain the behavior and silence this warning, please use dropout instead. Note that dropout2d exists to provide channel-wise dropout on inputs with 2 spatial dimensions, a channel dimension, and an optional batch dimension (i.e. 3D or 4D inputs).
```

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Minor comments below for warning wording to make deprecations actionable

File: torch/nn/functional.py [hunk 10]
```diff
@@ -1310,7 +1310,27 @@ def dropout2d(input: Tensor, p: float = 0.5, training: bool = True, inplace: boo
         return handle_torch_function(dropout2d, (input,), input, p=p, training=training, inplace=inplace)
     if p < 0.0 or p > 1.0:
         raise ValueError("dropout probability has to be between 0 and 1, " "but got {}".format(p))
-    return _VF.feature_dropout_(input, p, training) if inplace else _VF.feature_dropout(input, p, training)
+    inp_dim = input.dim()
+    msg = f"dropout2d: Input should be 3-D (unbatched) or 4-D (batched) but received {inp_dim}-D tensor"
+    assert inp_dim in (2, 3, 4), msg
+    if inp_dim == 2:
```
Comment (jbschlosser) [inline torch/nn/functional.py:1316]: @kshitij12345 Looks like some internal code is passing 5D inputs as well! How about this:
* Remove the assert entirely for now
* Throw the warning when encountering any input dim not in `(3, 4)`
* Analogous change for `Dropout3d`

Comment (kshitij12345) [reply torch/nn/functional.py]: Done.

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! One minor thing and we can get this in :)

File: torch/nn/functional.py [hunk 11]
```diff
@@ -1311,9 +1311,7 @@ def dropout2d(input: Tensor, p: float = 0.5, training: bool = True, inplace: boo
     if p < 0.0 or p > 1.0:
         raise ValueError("dropout probability has to be between 0 and 1, " "but got {}".format(p))
     inp_dim = input.dim()
-    msg = f"dropout2d: Input should be 3-D (unbatched) or 4-D (batched) but received {inp_dim}-D tensor"
-    assert inp_dim in (2, 3, 4), msg
-    if inp_dim == 2:
+    if inp_dim not in (3, 4):
         warn_msg = ("dropout2d: Received a 2D input to dropout2d, which is deprecated "
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/nn/functional.py]: Sorry. Thanks :)
